Sudden problems about device

### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.


### Question

It has been used normally for a long time, and today this problem suddenly appeared 。
-------------------------------------------------------------------------------------------------------------


Traceback (most recent call last):
  File "C:\Users\Administrator/.cache\torch\hub\ultralytics_yolov5_master\hubconf.py", line 43, in _create
    device = select_device(('0' if torch.cuda.is_available() else 'cpu') if device is None else device)
  File "C:\Users\Administrator/.cache\torch\hub\ultralytics_yolov5_master\utils\torch_utils.py", line 74, in select_device
    assert nd > int(max(device.split(','))), f'Invalid `--device {device}` request, valid devices are 0 - {nd - 1}'
AssertionError: Invalid `--device 0` request, valid devices are 0 - -1

### Additional

none

--- a/utils/datasets.py
+++ b/utils/datasets.py
@@ -29,13 +29,12 @@ from tqdm import tqdm
 from utils.augmentations import Albumentations, augment_hsv, copy_paste, letterbox, mixup, random_perspective
 from utils.general import (LOGGER, NUM_THREADS, check_dataset, check_requirements, check_yaml, clean_str,
                            segments2boxes, xyn2xy, xywh2xyxy, xywhn2xyxy, xyxy2xywhn)
-from utils.torch_utils import device_count, torch_distributed_zero_first
+from utils.torch_utils import torch_distributed_zero_first
 
 # Parameters
 HELP_URL = 'https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data'
 IMG_FORMATS = ['bmp', 'dng', 'jpeg', 'jpg', 'mpo', 'png', 'tif', 'tiff', 'webp']  # include image suffixes
 VID_FORMATS = ['asf', 'avi', 'gif', 'm4v', 'mkv', 'mov', 'mp4', 'mpeg', 'mpg', 'wmv']  # include video suffixes
-DEVICE_COUNT = max(device_count(), 1)  # number of CUDA devices
 
 # Get orientation exif tag
 for orientation in ExifTags.TAGS.keys():
@@ -110,7 +109,8 @@ def create_dataloader(path, imgsz, batch_size, stride, single_cls=False, hyp=Non
                                       prefix=prefix)
 
     batch_size = min(batch_size, len(dataset))
-    nw = min([os.cpu_count() // DEVICE_COUNT, batch_size if batch_size > 1 else 0, workers])  # number of workers
+    nd = torch.cuda.device_count()  # number of CUDA devices
+    nw = min([os.cpu_count() // max(nd, 1), batch_size if batch_size > 1 else 0, workers])  # number of workers
     sampler = None if rank == -1 else distributed.DistributedSampler(dataset, shuffle=shuffle)
     loader = DataLoader if image_weights else InfiniteDataLoader  # only DataLoader allows for attribute updates
     return loader(dataset,

--- a/utils/torch_utils.py
+++ b/utils/torch_utils.py
@@ -65,7 +65,6 @@ def device_count():
 
 
 def select_device(device='', batch_size=0, newline=True):
-    print(device)
     # device = 'cpu' or '0' or '0,1,2,3'
     s = f'YOLOv5 🚀 {git_describe() or date_modified()} torch {torch.__version__} '  # string
     device = str(device).strip().lower().replace('cuda:', '')  # to string, 'cuda:0' to '0'

--- a/utils/torch_utils.py
+++ b/utils/torch_utils.py
@@ -8,7 +8,6 @@ import math
 import os
 import platform
 import subprocess
-import sys
 import time
 from contextlib import contextmanager
 from copy import deepcopy
@@ -56,7 +55,7 @@ def git_describe(path=Path(__file__).parent):  # path must be a directory
 
 def device_count():
     # Returns number of CUDA devices available. Safe version of torch.cuda.device_count(). Only works on Linux.
-    assert sys.platform == 'Linux'
+    assert platform.system() == 'Linux'
     try:
         cmd = 'nvidia-smi -L | wc -l'
         return int(subprocess.run(cmd, shell=True, capture_output=True, check=True).stdout.decode().split()[-1])

--- a/utils/torch_utils.py
+++ b/utils/torch_utils.py
@@ -72,9 +72,8 @@ def select_device(device='', batch_size=0, newline=True):
         os.environ['CUDA_VISIBLE_DEVICES'] = '-1'  # force torch.cuda.is_available() = False
     elif device:  # non-cpu device requested
         os.environ['CUDA_VISIBLE_DEVICES'] = device  # set environment variable - must be before assert is_available()
-        nd = torch.cuda.device_count()  # number of CUDA devices
-        assert torch.cuda.is_available(), f'CUDA `--device {device}` requested but `torch.cuda.is_available()=False`'
-        assert nd > 0, f'Invalid CUDA `--device {device}` requested, use `--device cpu` or pass valid CUDA device(s)'
+        assert torch.cuda.is_available(), \
+            f"Invalid CUDA '--device {device}' requested, use '--device cpu' or pass valid CUDA device(s)"
 
     cuda = not cpu and torch.cuda.is_available()
     if cuda:

--- a/utils/torch_utils.py
+++ b/utils/torch_utils.py
@@ -72,7 +72,7 @@ def select_device(device='', batch_size=0, newline=True):
         os.environ['CUDA_VISIBLE_DEVICES'] = '-1'  # force torch.cuda.is_available() = False
     elif device:  # non-cpu device requested
         os.environ['CUDA_VISIBLE_DEVICES'] = device  # set environment variable - must be before assert is_available()
-        assert torch.cuda.is_available(), \
+        assert torch.cuda.is_available() and torch.cuda.device_count() >= len(device.replace(',', '')), \
             f"Invalid CUDA '--device {device}' requested, use '--device cpu' or pass valid CUDA device(s)"
 
     cuda = not cpu and torch.cuda.is_available()

--- a/utils/torch_utils.py
+++ b/utils/torch_utils.py
@@ -8,6 +8,7 @@ import math
 import os
 import platform
 import subprocess
+import sys
 import time
 from contextlib import contextmanager
 from copy import deepcopy
@@ -54,7 +55,8 @@ def git_describe(path=Path(__file__).parent):  # path must be a directory
 
 
 def device_count():
-    # Returns number of CUDA devices available. Safe version of torch.cuda.device_count().
+    # Returns number of CUDA devices available. Safe version of torch.cuda.device_count(). Only works on Linux.
+    assert sys.platform == 'Linux'
     try:
         cmd = 'nvidia-smi -L | wc -l'
         return int(subprocess.run(cmd, shell=True, capture_output=True, check=True).stdout.decode().split()[-1])
@@ -63,6 +65,7 @@ def device_count():
 
 
 def select_device(device='', batch_size=0, newline=True):
+    print(device)
     # device = 'cpu' or '0' or '0,1,2,3'
     s = f'YOLOv5 🚀 {git_describe() or date_modified()} torch {torch.__version__} '  # string
     device = str(device).strip().lower().replace('cuda:', '')  # to string, 'cuda:0' to '0'
@@ -70,10 +73,10 @@ def select_device(device='', batch_size=0, newline=True):
     if cpu:
         os.environ['CUDA_VISIBLE_DEVICES'] = '-1'  # force torch.cuda.is_available() = False
     elif device:  # non-cpu device requested
-        nd = device_count()  # number of CUDA devices
-        assert nd > int(max(device.split(','))), f'Invalid `--device {device}` request, valid devices are 0 - {nd - 1}'
         os.environ['CUDA_VISIBLE_DEVICES'] = device  # set environment variable - must be before assert is_available()
-        assert torch.cuda.is_available(), 'CUDA is not available, use `--device cpu` or do not pass a --device'
+        nd = torch.cuda.device_count()  # number of CUDA devices
+        assert torch.cuda.is_available(), f'CUDA `--device {device}` requested but `torch.cuda.is_available()=False`'
+        assert nd > 0, f'Invalid CUDA `--device {device}` requested, use `--device cpu` or pass valid CUDA device(s)'
 
     cuda = not cpu and torch.cuda.is_available()
     if cuda:

--- a/utils/torch_utils.py
+++ b/utils/torch_utils.py
@@ -55,7 +55,7 @@ def git_describe(path=Path(__file__).parent):  # path must be a directory
 
 def device_count():
     # Returns number of CUDA devices available. Safe version of torch.cuda.device_count(). Only works on Linux.
-    assert platform.system() == 'Linux'
+    assert platform.system() == 'Linux', 'device_count() function only works on Linux'
     try:
         cmd = 'nvidia-smi -L | wc -l'
         return int(subprocess.run(cmd, shell=True, capture_output=True, check=True).stdout.decode().split()[-1])

